tests/secret_watcher_test.go

```diff
@@ -0,0 +1,474 @@
+	}, 1*time.Minute, 5*time.Second, "password should be updated to the new value from the secret")
+
+	// get the final secret
+	finalSecret := &corev1.Secret{}
+	err = k8sClient.Get(ctx, types.NamespacedName{
+		Name:      user.Spec.ConnInfoSecretTarget.Name,
+		Namespace: user.Namespace,
+	}, finalSecret)
+	require.NoError(t, err)
+
+	assert.NoError(t, pingClickhouse(
+		ctx,
+		finalSecret.Data["HOST"],
+		finalSecret.Data["PORT"],
+		finalSecret.Data["USERNAME"],
+		finalSecret.Data["PASSWORD"],
+	), "updated password should allow ClickHouse connection")
+
+	// test that the OLD password no longer works
+	assert.Error(t, pingClickhouse(
+		ctx,
+		finalSecret.Data["HOST"],
+		finalSecret.Data["PORT"],
+		finalSecret.Data["USERNAME"],
+		[]byte("clickhouse-password-12345"),
+	), "old password should no longer work after update")
+}
+
+func getClickhouseUserWithSecretSourceYaml(project, serviceName, userName, secretName, cloudName string) string {
+	return fmt.Sprintf(`
+apiVersion: v1
+kind: Secret
+metadata:
+  name: %[4]s
+data:
+  PASSWORD: Y2xpY2tob3VzZS1wYXNzd29yZC0xMjM0NQ== # clickhouse-password-12345 # gitleaks:allow
+---
+apiVersion: aiven.io/v1alpha1
+kind: Clickhouse
+metadata:
+  name: %[2]s
+spec:
+  authSecretRef:
+    name: aiven-token
+    key: token
+
+  project: %[1]s
+  cloudName: %[5]s
+  plan: startup-16
+
+  userConfig:
+    public_access:
+      clickhouse: true
+---
+apiVersion: aiven.io/v1alpha1
+kind: ClickhouseUser
+metadata:
+  name: %[3]s
+spec:
+  authSecretRef:
+    name: aiven-token
+    key: token
+
+  connInfoSecretTarget:
+    name: my-clickhouse-user-secret
+
+  connInfoSecretSource:
+    name: %[4]s
+    passwordKey: PASSWORD
+
+  project: %[1]s
+  serviceName: %[2]s
+`, project, serviceName, userName, secretName, cloudName)
+}
+
+// TestCrossNamespaceSecretWatch tests secret watching across namespaces
+func TestCrossNamespaceSecretWatch(t *testing.T) {
+	defer recoverPanic(t)
+
+	// GIVEN
+	ctx, cancel := testCtx()
+	defer cancel()
+
+	serviceName := randName("cross-ns-pg")
+	userName := randName("cross-ns-user")
+	secretName := randName("cross-ns-secret")
+	secretNamespace := "test-secrets"
+
+	testNs := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: secretNamespace},
+	}
+	err := k8sClient.Create(ctx, testNs)
+	require.NoError(t, err)
+
+	defer func() {
+		deleteCtx, deleteCancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer deleteCancel()
+		_ = k8sClient.Delete(deleteCtx, testNs)
+	}()
+
+	yml := fmt.Sprintf(`
+apiVersion: v1
+kind: Secret
+metadata:
+  name: %[4]s
+  namespace: %[5]s
+data:
+  password: Y3Jvc3MtbnMtcGFzc3dvcmQtMTIzNDU= # cross-ns-password-12345 # gitleaks:allow
+---
+apiVersion: aiven.io/v1alpha1
+kind: PostgreSQL
+metadata:
+  name: %[2]s
+spec:
+  authSecretRef:
+    name: aiven-token
+    key: token
+
+  project: %[1]s
+  cloudName: %[6]s
+  plan: startup-16
+
+  userConfig:
+    pg_version: "15"
```

Comment:
Please remove the version: as far as I can tell it is not important for the test. Someday this one will reach the EOL and the test will fail.

Reply:
removed